_posts/2026-08-07-inside-a-kafka-order-workflow-outbox-retries-dlqs-and-idempotency.md

```
@@ -0,0 +1,358 @@
+Throughout this flow, all messages for the same order ID route to the same partition because you're using order ID as the key. You can scale the payment-service to multiple instances, and Kafka will distribute partitions among them. Each instance processes different orders in parallel, but events for any single order are always processed sequentially.
+
+```bash
+# Sample: Get current orders
+docker-compose exec order-service curl http://localhost:8080/orders
+
+# View topics & DLT events in Kafka UI
+```
+
+What you've built isn't just code—it's a system with explicit guarantees and well-defined failure modes. Orders won't disappear silently. Duplicate processing won't cause duplicate charges. Transient failures won't stall the entire system. Permanent failures won't create infinite retry loops. Each pattern addresses a specific class of failure, and together they form a resilient whole.
+
+![](/assets/img/uploads/untitled-diagram-2026-08-07-113500.png "Workflow diagram")
```

markdownlint-cli2 (0.23.2)

[warning] Images should have alternate text (alt text)

(MD045, no-alt-text)